File: frontend/lib/history.ts
import { invoke } from "@tauri-apps/api/core";

export interface CommandHistoryEntry {
  t: "cmd";
  v: number;
  ts: number;
  sid: string;
  c: string;
  exit: number;
  count: number;
}

export interface PromptHistoryEntry {
  t: "prompt";
  v: number;
  ts: number;
  sid: string;
  c: string;
  model: string;
  provider: string;
  tok_in: number;
  tok_out: number;
  ok: boolean;
  count: number;
}

export type HistoryEntry = CommandHistoryEntry | PromptHistoryEntry;

export async function addCommandHistory(
  sessionId: string,
  command: string,
  exitCode: number
): Promise<void> {
  return invoke("add_command_history", {
    sessionId,
    command,
    exitCode,
  });
}

export async function addPromptHistory(
  sessionId: string,
  prompt: string,
  model: string,
  provider: string,
  tokensIn: number,
  tokensOut: number,
  success: boolean
): Promise<void> {
  return invoke("add_prompt_history", {
    sessionId,
    prompt,
    model,
    provider,
    tokensIn,
    tokensOut,
    success,
  });
}

export async function loadHistory(
  limit: number = 500,
  entryType?: "cmd" | "prompt"
): Promise<HistoryEntry[]> {
  return invoke("load_history", { limit, entryType });
}

export async function searchHistory(
  query: string,
  includeArchives: boolean = false,
  limit: number = 200,
  entryType?: "cmd" | "prompt"
): Promise<HistoryEntry[]> {
  return invoke("search_history", { query, includeArchives, limit, entryType });
}

export async function clearHistory(): Promise<void> {
  return invoke("clear_history");
}

File: backend/crates/qbit/src/commands/history.rs
use crate::error::Result;
use crate::history::{HistoryEntry, HistoryManager};
use crate::state::AppState;
use std::sync::Arc;
use tauri::State;
use tokio::sync::RwLock;

#[tauri::command]
pub async fn add_command_history(
    _state: State<'_, AppState>,
    history: State<'_, Arc<RwLock<Option<HistoryManager>>>>,
    session_id: String,
    command: String,
    exit_code: i32,
) -> Result<()> {
    if let Some(history) = history.read().await.as_ref() {
        history
            .add_command(session_id, command, exit_code)
            .map_err(|e| crate::error::QbitError::Internal(e.to_string()))?;
    }
    Ok(())
}

#[allow(clippy::too_many_arguments)]
#[tauri::command]
pub async fn add_prompt_history(
    _state: State<'_, AppState>,
    history: State<'_, Arc<RwLock<Option<HistoryManager>>>>,
    session_id: String,
    prompt: String,
    model: String,
    provider: String,
    tokens_in: u32,
    tokens_out: u32,
    success: bool,
) -> Result<()> {
    if let Some(history) = history.read().await.as_ref() {
        history
            .add_prompt(
                session_id, prompt, model, provider, tokens_in, tokens_out, success,
            )
            .map_err(|e| crate::error::QbitError::Internal(e.to_string()))?;
    }
    Ok(())
}

#[tauri::command]
pub async fn load_history(
    _state: State<'_, AppState>,
    history: State<'_, Arc<RwLock<Option<HistoryManager>>>>,
    limit: usize,
    entry_type: Option<String>,
) -> std::result::Result<Vec<HistoryEntry>, String> {
    let Some(ref history) = *history.read().await else {
        return Ok(vec![]);
    };
    let et = entry_type.as_deref();
    history.load_recent(limit, et).map_err(|e| e.to_string())
}

#[tauri::command]
pub async fn search_history(
    _state: State<'_, AppState>,
    history: State<'_, Arc<RwLock<Option<HistoryManager>>>>,
    query: String,
    include_archives: bool,
    limit: usize,
    entry_type: Option<String>,
) -> std::result::Result<Vec<HistoryEntry>, String> {
    let Some(ref history) = *history.read().await else {
        return Ok(vec![]);
    };
    let et = entry_type.as_deref();
    history
        .search(query, include_archives, limit, et)
        .map_err(|e| e.to_string())
}

#[tauri::command]
pub async fn clear_history(
    _state: State<'_, AppState>,
    history: State<'_, Arc<RwLock<Option<HistoryManager>>>>,
) -> std::result::Result<(), String> {
    let Some(ref history) = *history.read().await else {
        return Ok(());
    };
    history.clear_all().map_err(|e| e.to_string())
}
